src/layers/ordered.rs

```diff
@@ -0,0 +1,514 @@
+    }
+
+    fn num_entries_deep(&self) -> usize {
+        self.tuples()
+    }
+
+    fn const_num_entries() -> Option<usize> {
+        None
+    }
+}
+
+impl<K, L, O> SharedRef for OrderedLayer<K, L, O>
+where
+    K: Ord + Clone,
+    L: Clone,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    type Target = Self;
+
+    fn try_into_owned(self) -> Result<Self::Target, Self> {
+        Ok(self)
+    }
+}
+
+impl<K, L, O> Trie for OrderedLayer<K, L, O>
+where
+    K: Ord + Clone,
+    L: Trie,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    type Key = K;
+    type ChildKey = L::Key;
+    type Item = (K, L::Item);
+    type Cursor = OrderedCursor<K, L, O>;
+    type MergeBuilder = OrderedBuilder<K, L::MergeBuilder, O>;
+    type TupleBuilder = UnorderedBuilder<K, L::TupleBuilder, O>;
+
+    fn keys(&self) -> usize {
+        self.keys.len()
+    }
+    fn tuples(&self) -> usize {
+        self.vals.tuples()
+    }
+    fn cursor_from(&self, lower: usize, upper: usize) -> Self::Cursor {
+        if lower < upper {
+            OrderedCursor {
+                bounds: (lower, upper),
+                pos: lower,
+                _types: PhantomData,
+            }
+        } else {
+            OrderedCursor {
+                bounds: (0, 0),
+                pos: 0,
+                _types: PhantomData,
+            }
+        }
+    }
+}
+
+/// Assembles a layer of this
+pub struct OrderedBuilder<K, L, O = usize>
+where
+    K: Ord,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    /// Keys
+    pub keys: Vec<K>,
+    /// Offsets
+    pub offs: Vec<O>,
+    /// The next layer down
+    pub vals: L,
+}
+
+impl<K, L, O> Builder for OrderedBuilder<K, L, O>
+where
+    K: Ord + Clone,
+    L: Builder,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    type Trie = OrderedLayer<K, L::Trie, O>;
+    fn boundary(&mut self) -> usize {
+        self.offs[self.keys.len()] = O::try_from(self.vals.boundary()).unwrap();
+        self.keys.len()
+    }
+    fn done(mut self) -> Self::Trie {
+        if !self.keys.is_empty() && self.offs[self.keys.len()].try_into().unwrap() == 0 {
+            self.offs[self.keys.len()] = O::try_from(self.vals.boundary()).unwrap();
+        }
+        OrderedLayer {
+            keys: self.keys,
+            offs: self.offs,
+            vals: self.vals.done(),
+        }
+    }
+}
+
+impl<K, L, O> MergeBuilder for OrderedBuilder<K, L, O>
+where
+    K: Ord + Clone,
+    L: MergeBuilder,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    fn with_capacity(keys: usize, tuples: usize) -> Self {
+        let mut offs = Vec::with_capacity(keys + 1);
+        offs.push(O::try_from(0_usize).unwrap());
+        OrderedBuilder {
```

Comment:
Another place for `OrdOffset::zero()` too
```suggestion
        offs.push(O::try_from(0_usize).unwrap());

        OrderedBuilder {
```